controlplane/segment_allocator.h

```diff
@@ -0,0 +1,308 @@
+		}
+		uint32_t result = size_info.head_block * BlockSize + index_in_block * size + IndexBegin;
+		if (block.free_segments == 0)
+		{
+			// block is busy
+			RemoveFromList(size_info.head_block, size);
+			size_info.busy_blocks++;
+		}
+		size_info.used_segments++;
+
+		free_cells_ -= size;
+		return result;
+	}
+
+	uint32_t Free(uint32_t start, uint16_t size)
+	{
+		if ((size == 0) || (size > MaxBufferSize) || (start >= IndexEnd) || (start < IndexBegin))
+		{
+			errors_external_++;
+			return false;
+		}
+
+		uint32_t block_index = (start - IndexBegin) / BlockSize;
+		OneBlock& block = all_blocks_[block_index];
+		if (block.one_segment_size != size)
+		{
+			errors_external_++;
+			return false;
+		}
+		else if ((start - block_index * BlockSize - IndexBegin) % size != 0)
+		{
+			errors_external_++;
+			return false;
+		}
+
+		bool block_was_busy = (block.free_segments == 0);
+		uint16_t index = (start - block_index * BlockSize - IndexBegin) / size;
+		if (block.Free(index) == error_in_block_)
+		{
+			errors_external_++;
+			return false;
+		}
+
+		OneSizeBlockInfo& size_info = sizes_info_[size];
+		if (block_was_busy)
+		{
+			if ((block.previous != null_block_) || (block.next != null_block_))
+			{
+				errors_internal_++;
+			}
+			InsertToList(block_index, size);
+			size_info.busy_blocks--;
+		}
+		if (block.free_segments == block.total_segments)
+		{
+			RemoveFromList(block_index, size);
+			InsertToList(block_index, 0);
+		}
+
+		size_info.used_segments--;
+		free_cells_ += size;
+		return true;
+	}
+
+	uint32_t Size() const
+	{
+		return free_cells_;
+	}
+
+	const OneSizeBlockInfo* GetBlocksStat() const
+	{
+		return sizes_info_;
+	}
+
+	std::pair<uint64_t, uint64_t> GetErrors() const
+	{
+		return {errors_external_, errors_internal_};
+	}
+
+private:
+	struct OneBlock
+	{
+		// indexes of previous and next blocks in list
+		uint32_t previous;
+		uint32_t next;
+
+		// block structure
+		uint16_t block_size;
+		uint16_t one_segment_size;
+		uint16_t total_segments;
+		uint16_t free_segments;
+
+		// bit masks of segments usage
+		uint64_t group_mask;
+		uint64_t masks[BlockSize / 64];
+
+		void Initialize(uint32_t previous, uint32_t next)
+		{
+			this->previous = previous;
+			this->next = next;
+			block_size = BlockSize;
+			group_mask = 0;
+			memset(masks, 0, sizeof(masks));
+		}
+
+		void SetSize(uint16_t segment_size)
+		{
+			one_segment_size = segment_size;
+			total_segments = block_size / segment_size;
+			free_segments = total_segments;
+		}
+
+		uint16_t Allocate()
+		{
+			// calling the function implies that we are sure that there are free segments
+			if ((~group_mask) == 0)
+			{
+				return error_in_block_;
+			}
+
+			uint16_t group = __builtin_ctzll(~group_mask);
+			if ((group >= sizeof(masks) / sizeof(masks[0])) || ((~masks[group]) == 0))
+			{
+				return error_in_block_;
+			}
+			uint16_t index_in_group = __builtin_ctzll(~masks[group]);
+			uint16_t index = group * 64 + index_in_group;
```

Comment:
```cpp
            auto group = static_cast<uint16_t>(group_mask._Find_first());
            if (group >= masks.size() || masks[group].all())
            {
                return std::nullopt;
            }
            auto index_in_group = static_cast<uint16_t>(masks[group]._Find_first());
            uint16_t index = group * BITS_PER_GROUP + index_in_group;
```

`_Find_first` will use `__builtin_ctzll` so there will be no overhead, if that's a concern